conan/tools/env/environment.py

```diff
@@ -0,0 +1,265 @@
+class _EnvVarPlaceHolder:
+    pass
+
+
+class _Sep(str):
+    pass
+
+
+class _PathSep:
+    pass
+
+
+def environment_wrap_command(filename, cmd):
+    if filename.endswith(".bat"):
+        return "{} && {}".format(filename, cmd)
+    elif filename.endswith(".sh"):
+        return 'bash -c ". {} && {}"'.format(filename, cmd.replace('"', r'\"'))
+    raise Exception("Unsupported environment file type {}".format(filename))
+
+
+class Environment:
+    def __init__(self):
+        # TODO: Maybe we need to pass conanfile to get the [conf]
+        # It being ordered allows for Windows case-insensitive composition
+        self._values = OrderedDict()  # {var_name: [] of values, including separators}
+
+    def __repr__(self):
+        return repr(self._values)
+
+    def vars(self):
+        return list(self._values.keys())
+
+    def value(self, name, placeholder="{name}", pathsep=os.pathsep):
+        return self._format_value(name, self._values[name], placeholder, pathsep)
+
+    @staticmethod
+    def _format_value(name, varvalues, placeholder, pathsep):
+        values = []
+        for v in varvalues:
+
+            if v is _EnvVarPlaceHolder:
+                values.append(placeholder.format(name=name))
+            elif v is _PathSep:
+                values.append(pathsep)
+            else:
+                values.append(v)
+        return "".join(values)
+
+    @staticmethod
+    def _list_value(value, separator):
+        if isinstance(value, list):
+            result = []
+            for v in value[:-1]:
+                result.append(v)
+                result.append(separator)
+            result.extend(value[-1:])
+            return result
+        else:
+            return [value]
+
+    def define(self, name, value, separator=" "):
+        value = self._list_value(value, _Sep(separator))
+        self._values[name] = value
+
+    def define_path(self, name, value):
+        value = self._list_value(value, _PathSep)
+        self._values[name] = value
+
+    def unset(self, name):
+        """
+        clears the variable, equivalent to a unset or set XXX=
+        """
+        self._values[name] = []
+
+    def append(self, name, value, separator=" "):
+        value = self._list_value(value, _Sep(separator))
+        self._values[name] = [_EnvVarPlaceHolder] + [_Sep(separator)] + value
+
+    def append_path(self, name, value):
+        value = self._list_value(value, _PathSep)
+        self._values[name] = [_EnvVarPlaceHolder] + [_PathSep] + value
+
+    def prepend(self, name, value, separator=" "):
+        value = self._list_value(value, _Sep(separator))
+        self._values[name] = value + [_Sep(separator)] + [_EnvVarPlaceHolder]
+
+    def prepend_path(self, name, value):
+        value = self._list_value(value, _PathSep)
+        self._values[name] = value + [_PathSep] + [_EnvVarPlaceHolder]
+
+    def save_bat(self, filename, generate_deactivate=True, pathsep=os.pathsep):
+        deactivate = textwrap.dedent("""\
```

Comment:
Yes, I agree it would be great to provide templates for this, and provide user extensibility. At this point my priority is to define the new paradigm:

- How generators access dependencies information via a visitor (ConanFileDependencies)
- The new separation of build and run envs
- The explicit generation of environment scripts for handling env-vars instead of doing everything dynamically in Conan (and thus not allowing many user flows)
- A clear and explicit mechanism to define env-var operations: append, prepend, clear, define, and how these compose for multiple levels (profile, build-requires, toolchains...)

Still the problem of .env files is not fully clear. In Conan we clearly have the user demand for those different operations: append, prepend, define, unset, and those .env files cannot do that. Also, in Conan users might need to concatenate more than one environment definition (for example the definition coming from profiles and build-requires, and another definition coming from a toolchain), that both will need to perform different operations like appending. Those .env files cannot do this either, up to my knowledge.